core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java

```diff
@@ -63,10 +60,9 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                 // skip
             } else if (token == XContentParser.Token.START_OBJECT) {
                 if ("query".equals(currentFieldName)) {
-                    query = parseContext.parseInnerQuery();
+                    query = parseContext.parseInnerQueryBuilder();
                 } else if ("filter".equals(currentFieldName)) {
-                    filterFound = true;
-                    filter = parseContext.parseInnerFilter();
+                    filter = parseContext.parseInnerQueryBuilder();
```

Comment:
This needs to be another method (`parseInnerFilterToQueryBuilder`) which replaces `parseInnerFilter` and also takes care of switching the interal `isFilter` flag.